test-harness/src/matcher/and_matcher.rs

```diff
@@ -0,0 +1,72 @@
+use std::fmt::Display;
+
+use crate::Event;
+use crate::Matcher;
+
+/// A [`Matcher`] that can match either of two other matchers.
+pub struct AndMatcher<A, B>(pub A, pub B);
+
+impl<A, B> Display for AndMatcher<A, B>
+where
+    A: Display,
+    B: Display,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{} and {}", self.0, self.1)
+    }
+}
+
+impl<A, B> Matcher for AndMatcher<A, B>
+where
+    A: Display + Matcher,
+    B: Display + Matcher,
+{
+    fn matches(&mut self, event: &Event) -> miette::Result<bool> {
+        // There may be some overlap in the events these matchers require to complete, so
+        // we eagerly evaluate both matchers before combining the boolean result.
+        let match_a = self.0.matches(event)?;
+        let match_b = self.1.matches(event)?;
+        Ok(match_a && match_b)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use tracing::Level;
+
+    use crate::tracing_json::Span;
+    use crate::IntoMatcher;
+
+    use super::*;
+
+    #[test]
+    fn test_and_matcher() {
+        let mut matcher = "puppy"
+            .into_matcher()
+            .unwrap()
+            .and("doggy".into_matcher().unwrap());
+        let event = Event {
+            message: "puppy".to_owned(),
+            timestamp: "2023-08-25T22:14:30.067641Z".to_owned(),
+            level: Level::INFO,
+            fields: Default::default(),
+            target: "ghciwatch::ghci".to_owned(),
+            span: Some(Span {
+                name: "ghci".to_owned(),
+                rest: Default::default(),
+            }),
+            spans: vec![Span {
+                name: "ghci".to_owned(),
+                rest: Default::default(),
+            }],
+        };
+
+        assert!(!matcher.matches(&event).unwrap());
+        assert!(matcher
+            .matches(&Event {
+                message: "doggy".to_owned(),
+                ..event
+            })
+            .unwrap());
```

Comment:
Feels strange to me that `matches` is mutating the matcher. I feel like consuming events and mutating state should be separate from checking whether the matcher is satisfied yet.

Something like this (dumb names aside):

```rust
matcher.consume(&event).unwrap();
assert!(!matcher.is_satisfied());

matcher.consume(&Event { message: "doggy".to_owned(), ..event }).unwrap();
assert!(matcher.is_satisfied());
```

Where `consume` takes a `&mut self` and `is_satisfied` takes a `&self`.

Reply:
Hm. My thought process is that the only thing that can update whether the matcher is satisfied is feeding it an event. Splitting it up like this would also mean storing some extra state, like `FusedMatcher` does.

This is a parallel to the [`Iterator::next`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#tymethod.next) method:

> Returns [`None`](https://doc.rust-lang.org/stable/std/option/enum.Option.html#variant.None) when iteration is finished. Individual iterator implementations may choose to resume iteration, and so calling `next()` again may or may not eventually start returning [`Some(Item)`](https://doc.rust-lang.org/stable/std/option/enum.Option.html#variant.Some) again at some point.

To compensate for this, there's a [`FusedIterator`](https://doc.rust-lang.org/stable/std/iter/trait.FusedIterator.html) trait:

> An iterator that always continues to yield None when exhausted.
>
> Calling `next` on a fused iterator that has returned `None` once is guaranteed to return [`None`](https://doc.rust-lang.org/stable/std/option/enum.Option.html#variant.None) again. This trait should be implemented by all iterators that behave this way because it allows optimizing [`Iterator::fuse()`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.fuse).

Reply:
> I'm struggling to see the connection to Rust's iterators here (mayyybe it feels kinda like a peekable iterator?)

You're correct, [`Peekable`](https://doc.rust-lang.org/stable/std/iter/struct.Peekable.html) is similar, except instead of storing if the iterator has ever returned `None`, it stores the iterator's next item. 

For some `Matcher`s (like the `BaseMatcher`), the matching is stateless and pure; if we wanted to have separate `consume` and `is_satisfied` methods, `BaseMatcher` would need to store an `is_satisfied: boolean` field to remember if it has already matched. `FusedMatcher` behaves like this, storing an extra `matched: boolean` field and returning `Ok(true)` from `Matcher::matches` unconditionally if `matched` is `true`.

Similarly, `Iterator`'s `next()` method mutates the iterator's state to (possibly) provide a next element. Because `Iterator` doesn't have an `is_finished` method, returning `None` from `next()` once doesn't mean it will never return `Some` again (e.g. for an `Iterator` reading lines from a file, while some other process writes data to the file). So `std` provides a `FusedIterator` that stores an extra bit of data: has the underlying iterator returned `None` yet? If it has, `Iterator::next` always returns `None`, and otherwise it calls the underlying method.

It's the same principle, adding an extra bit of state to make another guarantee on the API.

Reply:
The underlying reason for `Peekable` is the same, too -- sometimes it's easy to check if there's a next element in advance, sometimes it's not, but it's always easy to take an unpeekable iterator, call `next` on it, and memoize the return value.

Reply:
Ah okay, I see how it's similar to `Iterator` now, thanks for explaining.